README.md
# sofa-rpc-node

[SOFARPC](https://github.com/sofastack/sofa-rpc) Nodejs 实现版本

[![NPM version][npm-image]][npm-url]
[![Node.js CI](https://github.com/sofastack/sofa-rpc-node/actions/workflows/nodejs.yml/badge.svg)](https://github.com/sofastack/sofa-rpc-node/actions/workflows/nodejs.yml)
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/sofa-rpc-node.svg?style=flat-square
[npm-url]: https://npmjs.org/package/sofa-rpc-node
[snyk-image]: https://snyk.io/test/npm/sofa-rpc-node/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/sofa-rpc-node
[download-image]: https://img.shields.io/npm/dm/sofa-rpc-node.svg?style=flat-square
[download-url]: https://npmjs.org/package/sofa-rpc-node

## 一、SOFARPC Node 简介

简单说它是 [SOFARPC](https://github.com/sofastack/sofa-rpc) 的 Nodejs 版实现，但本质上它是一个通用的 Nodejs RPC 解决方案。Nodejs RPC 在阿里和蚂蚁内部已经发展了四五年时间，如今广泛应用于各类业务场景，并经历了多次双 11 大促的考验。功能方面从基本的服务发布、寻址、点对点远程调用能力；到各种路由、负载均衡策略；再到故障隔离、熔断等高级功能，已逐渐发展成一个高可扩展性、高性能、生产级的 RPC 框架。

## 二、模块划分

SOFARPC Node 主要包含了四个子模块，分别是：

- __client：__ RPC 的客户端实现
- __server：__ RPC 的服务端实现
- __registry：__ 服务注册中心抽象及实现（目前提供 zookeeper 实现）
- __test：__ RPC 测试工具类

```bash
.
└── lib
    ├── client
    ├── registry
    ├── server
    └── test
```

## 三、快速上手

### 安装

```bash
npm install sofa-rpc-node --save
```

### 安装并启动 zookeeper

sofa-rpc-node 默认的注册中心实现基于 zookeeper，所以需要先启动一个 zookeeper 实例

从 Homebrew 安装（macOs）

```bash
brew install zookeeper
```

启动 zk server（默认端口为 2181）

```bash
zkServer start

# ZooKeeper JMX enabled by default
# Using config: /usr/local/etc/zookeeper/zoo.cfg
# Starting zookeeper ... STARTED
```

### 代码示例

- 暴露一个 RPC 服务，并发布到注册中心

```js
const { RpcServer } = require('sofa-rpc-node').server;
const { ZookeeperRegistry } = require('sofa-rpc-node').registry;
const logger = console;

// 1. 创建 zk 注册中心客户端
const registry = new ZookeeperRegistry({
  logger,
  address: '127.0.0.1:2181', // 需要本地启动一个 zkServer
});

// 2. 创建 RPC Server 实例
const server = new RpcServer({
  logger,
  registry, // 传入注册中心客户端
  port: 12200,
});

// 3. 添加服务
server.addService({
  interfaceName: 'com.nodejs.test.TestService',
}, {
  async plus(a, b) {
    return a + b;
  },
});

// 4. 启动 Server 并发布服务
server.start()
  .then(() => {
    server.publish();
  });
```

- 调用 RPC 服务（从注册中心获取服务列表）

```js
const { RpcClient } = require('sofa-rpc-node').client;
const { ZookeeperRegistry } = require('sofa-rpc-node').registry;
const logger = console;

// 1. 创建 zk 注册中心客户端
const registry = new ZookeeperRegistry({
  logger,
  address: '127.0.0.1:2181',
});

async function invoke() {
  // 2. 创建 RPC Client 实例
  const client = new RpcClient({
    logger,
    registry,
  });
  // 3. 创建服务的 consumer
  const consumer = client.createConsumer({
    interfaceName: 'com.nodejs.test.TestService',
  });
  // 4. 等待 consumer ready（从注册中心订阅服务列表...）
  await consumer.ready();

  // 5. 执行泛化调用
  const result = await consumer.invoke('plus', [ 1, 2 ], { responseTimeout: 3000 });
  console.log('1 + 2 = ' + result);
}

invoke().catch(console.error);
```

- 调用 RPC 服务（直连模式）

```js
const { RpcClient } = require('sofa-rpc-node').client;
const logger = console;

async function invoke() {
  // 不需要传入 registry 实例了
  const client = new RpcClient({
    logger,
  });
  const consumer = client.createConsumer({
    interfaceName: 'com.nodejs.test.TestService',
    serverHost: '127.0.0.1:12200', // 直接指定服务地址
  });
  await consumer.ready();

  const result = await consumer.invoke('plus', [ 1, 2 ], { responseTimeout: 3000 });
  console.log('1 + 2 = ' + result);
}

invoke().catch(console.error);
```

- 测试 RPC Server 的方法（用于单元测试）

```js
const request = require('sofa-rpc-node').test;
const { RpcServer } = require('sofa-rpc-node').server;
const logger = console;

describe('test/server.test.js', () => {
  let server;
  before(async function() {
    server = new RpcServer({
      logger,
      port: 12200,
    });
    server.addService({
      interfaceName: 'com.nodejs.test.TestService',
    }, {
      async plus(a, b) {
        return a + b;
      },
    });
    await server.start();
  });
  after(async function() {
    await server.close();
  });

  it('should call plus ok', async function() {
    await request(server)
      .service('com.nodejs.test.TestService')
      .invoke('plus')
      .send([ 1, 2 ])
      .expect(3);
  });
});
```

- 暴露和调用 protobuf 接口

通过 *.proto 来定义接口

```proto
syntax = "proto3";

package com.alipay.sofa.rpc.test;

// 可选
option java_multiple_files = false;

service ProtoService {
  rpc echoObj (EchoRequest) returns (EchoResponse) {}
}

message EchoRequest {
  string name = 1;
  Group group = 2;
}

message EchoResponse {
  int32 code = 1;
  string message = 2;
}

enum Group {
  A = 0;
  B = 1;
}
```

服务端代码

```js
const antpb = require('antpb');
const protocol = require('sofa-bolt-node');
const { RpcServer } = require('sofa-rpc-node').server;
const { ZookeeperRegistry } = require('sofa-rpc-node').registry;
const logger = console;

// 传入 *.proto 文件存放的目录，加载接口定义
const proto = antpb.loadAll('/path/proto');
// 将 proto 设置到协议中
protocol.setOptions({ proto });

const registry = new ZookeeperRegistry({
  logger,
  address: '127.0.0.1:2181',
});

const server = new RpcServer({
  logger,
  protocol, // 覆盖协议
  registry,
  codecType: 'protobuf', // 设置默认的序列化方式为 protobuf
  port: 12200,
});

server.addService({
  interfaceName: 'com.alipay.sofa.rpc.test.ProtoService',
}, {
  async echoObj(req) {
    req = req.toObject({ enums: String });
    return {
      code: 200,
      message: 'hello ' + req.name + ', you are in ' + req.group,
    };
  },
});
server.start()
  .then(() => {
    server.publish();
  });
```

客户端代码

```js
const antpb = require('antpb');
const protocol = require('sofa-bolt-node');
const { RpcClient } = require('sofa-rpc-node').client;
const { ZookeeperRegistry } = require('sofa-rpc-node').registry;
const logger = console;

// 传入 *.proto 文件存放的目录，加载接口定义
const proto = antpb.loadAll('/path/proto');
// 将 proto 设置到协议中
protocol.setOptions({ proto });

const registry = new ZookeeperRegistry({
  logger,
  address: '127.0.0.1:2181',
});

async function invoke() {
  const client = new RpcClient({
    logger,
    protocol,
    registry,
  });
  const consumer = client.createConsumer({
    interfaceName: 'com.alipay.sofa.rpc.test.ProtoService',
  });
  await consumer.ready();

  const result = await consumer.invoke('echoObj', [{
    name: 'gxcsoccer',
    group: 'B',
  }], { responseTimeout: 3000 });
  console.log(result);
}

invoke().catch(console.error);
```

### 最佳实践

虽然上面我们提供了示例代码，但是我们并不推荐您直接使用该模块，因为它的定位是 RPC 基础模块，只提供基本的 API，对于业务开发者可能并不是非常友好。我们的最佳实践是通过插件将 RPC 能力集成到 [Egg.js](https://github.com/eggjs/egg) 框架里，提供更加直观友好的用户接口，让您就像使用本地方法一样使用 RPC。这块也会在近期开放，敬请期待！


## 三、相关文档

- [聊聊 Node.js RPC（一）— 协议](https://www.yuque.com/egg/nodejs/dklip5)
- [聊聊 Node.js RPC（二）— 服务发现](https://www.yuque.com/egg/nodejs/mhgl9f)

## 四、如何贡献

请告知我们可以为你做些什么，不过在此之前，请检查一下是否有已经[存在的Bug或者意见](https://github.com/sofastack/sofa-rpc-node/issues)。

如果你是一个代码贡献者，请参考[代码贡献规范](https://github.com/eggjs/egg/blob/master/CONTRIBUTING.zh-CN.md)。

## 五、开源协议

[MIT](LICENSE)


History.md

2.7.0 / 2022-12-05
==================

**features**
  * [[`178cb61`](http://github.com/sofastack/sofa-rpc-node/commit/178cb6168ba948a0899c0f9817fc161544a78932)] - feat(lib): support consumer ready timeout (#108) (KenyeeCheung <<me@kenyee.cc>>)

2.6.0 / 2022-11-25
==================

**features**
  * [[`a11b04e`](http://github.com/sofastack/sofa-rpc-node/commit/a11b04eb493061ef684f65a381360efe6fc41219)] - 👌 IMPROVE: Try to use ctx.logger first (#106) (fengmk2 <<fengmk2@gmail.com>>)

**others**
  * [[`590be81`](http://github.com/sofastack/sofa-rpc-node/commit/590be8156112f96a343a6cb3a269f12a1831a93c)] - 🤖 TEST: Only test Node.js >= 14 (#105) (fengmk2 <<fengmk2@gmail.com>>)

2.5.0 / 2022-11-06
==================

**others**
  * [[`c32eb53`](http://github.com/sofastack/sofa-rpc-node/commit/c32eb53d24fbeb15b801c9ad25bb9bfe2fc3dc58)] - chore: add parseRequestPropsWithHeader decode option (#104) (mytEcust <<mytfromecust@gmail.com>>)

2.4.0 / 2022-09-26
==================

**features**
  * [[`a8201b7`](http://github.com/sofastack/sofa-rpc-node/commit/a8201b72ccbbf6b54a101dee93acc3b0e869694d)] - feat(graceful): support serverGracefulIgnoreCode (#103) (hyj1991 <<yeekwanvong@gmail.com>>)

**others**
  * [[`e74e280`](http://github.com/sofastack/sofa-rpc-node/commit/e74e280e24b3e1c9810675bff5140e0b467cd600)] - Create codeql-analysis.yml (fengmk2 <<fengmk2@gmail.com>>)

2.3.1 / 2022-09-06
==================

**fixes**
  * [[`50b8500`](http://github.com/sofastack/sofa-rpc-node/commit/50b8500c7b7edf41f07aacf023da255a2acf9804)] - fix: print err if encoder/decoder error (#101) (killa <<killa123@126.com>>)

2.3.0 / 2022-04-12
==================

**features**
  * [[`a88bf21`](http://github.com/sofastack/sofa-rpc-node/commit/a88bf21cdf99db807655a7e49a61d7d2ca768795)] - feat: do loadBalancer close when address_group close (mansonchor <<mansonchor1987@gmail.com>>)

2.2.2 / 2022-03-18
==================

**fixes**
  * [[`91760de`](http://github.com/sofastack/sofa-rpc-node/commit/91760de4fbe3a61ea85a5018e72fedff5e5bad61)] - fix: 去掉 err 上附带 req 信息，打印 error 的话可能存在安全内容泄漏 (#94) (mansonchor.github.com <<mansonchor1987@gmail.com>>)

2.2.1 / 2022-03-17
==================

**others**
  * [[`e06219e`](http://github.com/sofastack/sofa-rpc-node/commit/e06219e5479504d92964d2114a5a35c39db3d690)] - fix AddressGroup Ready may on 'next' event many times (#86) (羊鹿 <<general_up@163.com>>)

2.2.0 / 2022-03-13
==================

**features**
  * [[`394b0ff`](http://github.com/sofastack/sofa-rpc-node/commit/394b0ff246280b50e8d11ab46c09c353d54f40c8)] - feat: 运行覆盖 ConnectionManager 初始化链接异常消息 (#91) (fengmk2 <<fengmk2@gmail.com>>)

2.1.0 / 2022-03-12
==================

**fixes**
  * [[`11411f2`](http://github.com/sofastack/sofa-rpc-node/commit/11411f2a543d47fc4fe7681336c0ec4b99ea792c)] - fix: 统一使用 connectionManager 提供的 API (mansonchor.github.com <<mansonchor1987@gmail.com>>)
  * [[`6717e4f`](http://github.com/sofastack/sofa-rpc-node/commit/6717e4f9d980d95fcdd79ec232e0e1f154cfa9f9)] - fix: should await ready in invoke insteadof _invoke (#74) (killa <<killa123@126.com>>)
  * [[`99124c9`](http://github.com/sofastack/sofa-rpc-node/commit/99124c92fb90076df1d158c23772855cc73421c2)] - fix: create consumer with cache: false, will not override client._consumerCache map (#71) (mansonchor.github.com <<mansonchor@126.com>>)

**others**
  * [[`23cc741`](http://github.com/sofastack/sofa-rpc-node/commit/23cc7412b3eaf3a07bff854d61661c87f0c5da7e)] - test: skip unstable assert (fengmk2 <<fengmk2@gmail.com>>)
  * [[`3436650`](http://github.com/sofastack/sofa-rpc-node/commit/3436650a8201806f3f88bdaad517586cc07e1fc8)] - chore: remove autod (fengmk2 <<fengmk2@gmail.com>>)
  * [[`b92c60d`](http://github.com/sofastack/sofa-rpc-node/commit/b92c60d151c84172d4726e40f60b1c14f1b79faf)] - test: run ci on GitHub Action (fengmk2 <<fengmk2@gmail.com>>)
  * [[`ef9ffa8`](http://github.com/sofastack/sofa-rpc-node/commit/ef9ffa8ac7178580b4c7c31e4e697eebbcf79393)] - refactor: support child class overwrite connections mng (#75) (killa <<killa123@126.com>>)

2.0.0 / 2020-07-07
==================

**others**
  * [[`87b7683`](http://github.com/alipay/sofa-rpc-node/commit/87b7683ce93b44235e27c4d91d9f40629b174591)] - chore: upgrade sofa-bolt-node@2.0.1 (#70) (zōng yǔ <<gxcsoccer@users.noreply.github.com>>)

1.19.0 / 2020-07-02
==================

**fixes**
  * [[`578528f`](http://github.com/alipay/sofa-rpc-node/commit/578528f4d20fff09d3ab566f59ccd3039cc4ecaf)] - fix: no provider wrapError (#69) (陈峰 <<348018533@qq.com>>)

**others**
  * [[`1560625`](http://github.com/alipay/sofa-rpc-node/commit/1560625421a8b0f5da519e2b2ba6b8dcba7dd218)] - egg的rpc.server支持customMeta配置，用于支持自定义元数据； (#67) (ChangedenChan <<changeden520@gmail.com>>)

1.18.0 / 2020-03-23
==================

**features**
  * [[`92e881c`](http://github.com/sofastack/sofa-rpc-node/commit/92e881c9071a354a6cc0fe8896f98f23bd62d946)] - feat: consumer parseUrl 前过滤非字符串内容 (#66) (mansonchor.github.com <<mansonchor@126.com>>)

1.17.0 / 2020-03-12
==================

**features**
  * [[`6c16373`](http://github.com/sofastack/sofa-rpc-node/commit/6c16373f28c48a7e95169487a94dac2c95084bf2)] - feat: createConsumer support cache control (#65) (mansonchor.github.com <<mansonchor@126.com>>)

1.16.1 / 2020-02-28
==================

**fixes**
  * [[`d18dabd`](http://github.com/sofastack/sofa-rpc-node/commit/d18dabda79ee44207183472c440e9f3d4b08db82)] - fix: mw should have no effective with result code (#63) (killa <<killa123@126.com>>)

1.16.0 / 2019-12-24
==================

**features**
  * [[`a7e5cdb`](http://github.com/sofastack/sofa-rpc-node/commit/a7e5cdb500edf8557a1437c7edb038967f9be480)] - feat: support disable client decode cache (#62) (killa <<killa123@126.com>>)

1.15.0 / 2019-12-24
==================

**features**
  * [[`ded6ad1`](http://github.com/sofastack/sofa-rpc-node/commit/ded6ad105a84a84929c999f5ee9e58433d518d2d)] - feat: support disable decode cache (#61) (killa <<killa123@126.com>>)

1.14.2 / 2019-10-16

example/pb_server.js
'use strict';

const path = require('path');
const antpb = require('antpb');
const protocol = require('sofa-bolt-node');
const { RpcServer } = require('../').server;
const { ZookeeperRegistry } = require('../').registry;
const logger = console;

const proto = antpb.loadAll(path.join(__dirname, '../test/fixtures/proto'));
protocol.setOptions({ proto });

const registry = new ZookeeperRegistry({
  logger,
  address: '127.0.0.1:2181',
});

const server = new RpcServer({
  logger,
  protocol,
  registry,
  codecType: 'protobuf',
  port: 12200,
});

server.addService({
  interfaceName: 'com.alipay.sofa.rpc.test.ProtoService',
}, {
  async echoObj(req) {
    req = req.toObject({ enums: String });
    return {
      code: 200,
      message: 'hello ' + req.name + ', you are in ' + req.group,
    };
  },
});
server.start()
  .then(() => {
    server.publish();
  });
